--- a/drivers/firmware/efi/efi.c	2016-12-12 22:04:43.262163972 +0000
+++ b/drivers/firmware/efi/efi.c	2016-12-12 22:05:20.405497070 +0000
@@ -383,7 +383,7 @@
 			return 0;
 		}
 	}
-	pr_err_once("requested map not found.\n");
+	//pr_err_once("requested map not found.\n");
 	return -ENOENT;
 }
 
--- a/drivers/firmware/efi/esrt.c	2016-12-12 22:03:59.222164251 +0000
+++ b/drivers/firmware/efi/esrt.c	2016-12-12 22:04:16.225497478 +0000
@@ -254,7 +254,7 @@
 
 	rc = efi_mem_desc_lookup(efi.esrt, &md);
 	if (rc < 0) {
-		pr_err("ESRT header is not in the memory map.\n");
+		//pr_err("ESRT header is not in the memory map.\n");
 		return;
 	}
 
